codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedStringGenerator.kt

```diff
@@ -145,39 +162,163 @@ class ConstrainedStringGenerator(
                 """
-                ##[derive(Debug, PartialEq)]
-                pub enum ${constraintViolation.name} {
-                    Length(usize),
+                fn check_length(string: &str) -> Result<(), $constraintViolation> {
+                    let length = string.chars().count();
+
+                    if $condition {
+                        Ok(())
+                    } else {
+                        Err($constraintViolation::Length(length))
+                    }
                 }
                 """,
             )
+        }
+    }
+}
 
-            if (shape.isReachableFromOperationInput()) {
-                rustBlock("impl ${constraintViolation.name}") {
-                    rustBlockTemplate(
-                        "pub(crate) fn as_validation_exception_field(self, path: #{String}) -> crate::model::ValidationExceptionField",
-                        "String" to RuntimeType.String,
-                    ) {
-                        rustBlock("match self") {
-                            rust(
-                                """
-                                Self::Length(length) => crate::model::ValidationExceptionField {
-                                    message: format!("${lengthTrait.validationErrorMessage()}", length, &path),
-                                    path,
-                                },
-                                """,
-                            )
-                        }
+private data class Pattern(val patternTrait: PatternTrait) : StringTraitInfo() {
+    override fun toTraitInfo(): TraitInfo {
+        val pattern = patternTrait.pattern
+
+        return TraitInfo(
+            { rust("let value = Self::check_pattern(value)?;") },
+            {
+                docs("Error when a string doesn't satisfy its `@pattern`.")
+                docs("Contains the String that failed the pattern.")
+                rust("Pattern(String),")
+            },
+            {
+                rust(
+                    """
+                    Self::Pattern(string) => crate::model::ValidationExceptionField {
+                        message: format!("${patternTrait.validationErrorMessage()}", &string, &path, r##"$pattern"##),
+                        path
+                    },
+                    """,
+                )
+            },
+            this::renderValidationFunction,
+        )
+    }
+
+    /**
+     * Renders a `check_pattern` function to validate the string matches the
+     * supplied regex in the `@pattern` trait.
+     */
+    private fun renderValidationFunction(constraintViolation: Symbol): Writable {
+        val pattern = patternTrait.pattern
+
+        return {
+            rustTemplate(
+                """
+                fn check_pattern(string: String) -> Result<String, $constraintViolation> {
+                    static REGEX : #{OnceCell}::sync::Lazy<#{Regex}::Regex> = #{OnceCell}::sync::Lazy::new(|| #{Regex}::Regex::new(r##"$pattern"##).unwrap());
```

Comment:
I thought I had left a comment here, but I can't find it in my first review.

Do you know if Smithy validates that the modeled regex is compliant with the ECMA 262 regular expression dialect, as per the spec? Or do they just accept any string as the value to the `pattern` trait?

If the former is true, if this line panics, it means that there are Smithy-compliant regexes that the `regex` crate does not accept. There's nothing we can do about it to recover though; I guess we could tell the user with `expect()` what's going on to help them tweak their modeled regex if possible, or file us an issue for support ([example](https://github.com/awslabs/smithy-rs/blob/995bf0f39043869372355b0538d640b1ce19733e/rust-runtime/aws-smithy-http-server/src/runtime_error.rs#L114-L114)).

If the latter is true, then that should probably be an issue on awslabs/smithy's side. Ideally they should fail at model build time.

In any case, I'm worried about the fact that we're compiling these regexes lazily for the sake of performance, and so it'll mean that the program won't panic until the service receives the first request for the operation whose closure contains a `pattern`-constrained shape. Until that happens, the server could have serviced a bunch of requests for other operations, and so a panic here instead of a graceful shutdown could result in things as bad as data corruption.

I think that ideally we should evaluate these eagerly, as soon as possible, and abort at service startup time, before the service has been set up. And carefully test against this failure scenario.

What do you think?

Reply:
Smithy currently [does validate the Regex](https://github.com/awslabs/smithy/blob/f53ffd52f0befdcb6ec5750bdb73f143bedeca4d/smithy-model/src/main/java/software/amazon/smithy/model/traits/PatternTrait.java#L57-L65), but it does so against [Java's regex library](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/regex/Pattern.html) which doesn't seem to be the exact same, but the differences are unclear to me.

**I've checked that Smithy does allow regexes that the `regex` crate doesn't support**, like this: `@pattern("import (?!static).+")` (which uses [negative look-ahead](https://www.baeldung.com/java-regex-lookahead-lookbehind#negative-lookahead)).

I'll look into eagerly compiling the regexes and failing early.

Reply:
I'll implement this in a follow-up PR. I've created a tracking issue here: https://github.com/awslabs/smithy-rs/issues/2026.